pandas/tests/indexes/test_base.py

```diff
@@ -237,61 +223,63 @@ def test_constructor_int_dtype_float(self, dtype):
-        tm.assert_index_equal(result, expected)
+        if klass_or_raises is True:
+            msg = "cannot convert"
+            with tm.assert_raises_regex(ValueError, msg):
+                Index(data, dtype=dtype)
+        else:
+            expected = klass_or_raises(data)
+            result = Index(data, dtype=dtype)
+            tm.assert_index_equal(result, expected)
 
-    def test_index_ctor_infer_nan_nat(self):
+    @pytest.mark.parametrize("klass,dtype,na_val", [
+        (pd.Float64Index, np.float64, np.nan),
+        (pd.DatetimeIndex, 'datetime64[ns]', pd.NaT)
+    ])
+    def test_index_ctor_infer_nan_nat(self, klass, dtype, na_val):
         # GH 13467
-        exp = pd.Float64Index([np.nan, np.nan])
-        assert exp.dtype == np.float64
-        tm.assert_index_equal(Index([np.nan, np.nan]), exp)
-        tm.assert_index_equal(Index(np.array([np.nan, np.nan])), exp)
-
-        exp = pd.DatetimeIndex([pd.NaT, pd.NaT])
-        assert exp.dtype == 'datetime64[ns]'
-        tm.assert_index_equal(Index([pd.NaT, pd.NaT]), exp)
-        tm.assert_index_equal(Index(np.array([pd.NaT, pd.NaT])), exp)
-
+        na_list = [na_val, na_val]
+        exp = klass(na_list)
+        assert exp.dtype == dtype
+        tm.assert_index_equal(Index(na_list), exp)
+        tm.assert_index_equal(Index(np.array(na_list)), exp)
+
+    @pytest.mark.parametrize("data", [
+        [pd.NaT, np.nan], [np.nan, pd.NaT], [np.nan, np.datetime64('nat')],
+        [np.datetime64('nat'), np.nan]
+    ])
+    def test_index_ctor_infer_nat_dti(self, data):
         exp = pd.DatetimeIndex([pd.NaT, pd.NaT])
         assert exp.dtype == 'datetime64[ns]'
 
-        for data in [[pd.NaT, np.nan], [np.nan, pd.NaT],
-                     [np.nan, np.datetime64('nat')],
-                     [np.datetime64('nat'), np.nan]]:
-            tm.assert_index_equal(Index(data), exp)
-            tm.assert_index_equal(Index(np.array(data, dtype=object)), exp)
+        tm.assert_index_equal(Index(data), exp)
+        tm.assert_index_equal(Index(np.array(data, dtype=object)), exp)
 
+    @pytest.mark.parametrize("data", [
+        [np.nan, np.timedelta64('nat')], [np.timedelta64('nat'), np.nan],
+        [pd.NaT, np.timedelta64('nat')], [np.timedelta64('nat'), pd.NaT]
+    ])
+    def test_index_ctor_infer_nat_tdi(self, data):
         exp = pd.TimedeltaIndex([pd.NaT, pd.NaT])
         assert exp.dtype == 'timedelta64[ns]'
 
-        for data in [[np.nan, np.timedelta64('nat')],
-                     [np.timedelta64('nat'), np.nan],
-                     [pd.NaT, np.timedelta64('nat')],
-                     [np.timedelta64('nat'), pd.NaT]]:
-            tm.assert_index_equal(Index(data), exp)
-            tm.assert_index_equal(Index(np.array(data, dtype=object)), exp)
+        tm.assert_index_equal(Index(data), exp)
+        tm.assert_index_equal(Index(np.array(data, dtype=object)), exp)
 
+    @pytest.mark.parametrize("swap_objs", [True, False])
```

Comment:
Kind of a strange parameter but just emulating the existing test - one assertion is done with the datetime at the initial position and another assertion is done with the timedelta at the initial position